# Import Libraries

In [1]:
import numpy as np
import pandas as pd
pd.set_option("max_columns", None)
from scipy.stats import chi2_contingency
from sklearn.linear_model import Lasso
from sklearn.linear_model import LassoCV
from sklearn.linear_model import Ridge
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

from keras.layers.normalization import BatchNormalization
from keras.layers.core import Activation, Dropout
from keras.optimizers import Adadelta
from keras.layers.advanced_activations import LeakyReLU
from keras.wrappers.scikit_learn import KerasRegressor
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.


# Import Data

In [2]:
data = pd.read_csv('fashion_data_on_Instagram.csv')
data.set_index('UserId', inplace = True)
data.head()

Followings  Followers  MediaCount  \
UserId                                                              
1171579752855683619_212070047        518.0     9840.0         541   
1171594777274371222_176762322       7333.0     2300.0         272   
1171407552643586413_581125501        131.0      605.0         106   
1171407552643586413_581125501        131.0      605.0         106   
1171508187966229230_2797323089       132.0      456.0          56   

                                  BrandName BrandCategory  \
UserId                                                      
1171579752855683619_212070047   abercrombie   High street   
1171594777274371222_176762322   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171508187966229230_2797323089  abercrombie   High street   

                                                                         Hashtags  \
UserId                                                                              
1171579752855683619_212070047   beautiful, summer, fashion, love, cute, food, ...   
1171594777274371222_176762322   teen, model, brunette, selfie, hollister, snap...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171508187966229230_2797323089  abercrombieandfitch, shopping, love, hollister...   

                                                                          Caption  \
UserId                                                                              
1171579752855683619_212070047     We were born to be REAL, not to be perfect. ...   
1171594777274371222_176762322     Gotta run, but first, let me take selfie. Me...   
1171407552643586413_581125501                         dress to impress\U0001f457    
1171407552643586413_581125501                         dress to impress\U0001f457    
1171508187966229230_2797323089    #love #shopping #shoppen #hollister #abercro...   

                                                                           ImgURL  \
UserId                                                                              
1171579752855683619_212070047   https://scontent.cdninstagram.com/t51.2885-15/...   
1171594777274371222_176762322   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171508187966229230_2797323089  https://scontent.cdninstagram.com/t51.2885-15/...   

                                Likes  Comments   CreationTime  \
UserId                                                           
1171579752855683619_212070047      97          0    1453883211   
1171594777274371222_176762322      94          0    1453885002   
1171407552643586413_581125501      91          3    1453862683   
1171407552643586413_581125501      94          3    1453862683   
1171508187966229230_2797323089      9          1    1453874680   

                                                                    Link  \
UserId                                                                     
1171579752855683619_212070047   https://www.instagram.com/p/BBCSZPzAMIj/   
1171594777274371222_176762322   https://www.instagram.com/p/BBCVz4YJEyW/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171508187966229230_2797323089  https://www.instagram.com/p/BBCCH1zMEru/   

                                  Selfie  BodySnap  Marketing  ProductOnly  \
UserId                                                                       
1171579752855683619_212070047   0.000885  0.336438   0.741220     0.009454   
1171594777274371222_176762322   0.003912  0.995503   0.001728     0.002125   
1171407552643586413_5811

# Clean Data

### Let's See if there are any NaNs in the data

In [16]:
hashtag = data.copy()
hashtag['Hashtags'] = data['Hashtags'].isna()
hashtag.groupby('Hashtags')['MediaCount'].count()

Hashtags
False    24606
True       146
Name: MediaCount, dtype: int64

In [17]:
Caption = data.copy()
Caption['Caption'] = data['Caption'].isna()
Caption.groupby('Caption')['MediaCount'].count()

Caption
False    24717
True        35
Name: MediaCount, dtype: int64

In [5]:
set(data['Comments '].isna())

{False}

##### So there exist NaN values in Hashtags and Captions. There are no other columns that contain NaN.

In [6]:
data.isna().groupby('Hashtags')['Followings'].count()

Hashtags
False    24606
True       146
Name: Followings, dtype: int64

##### There are 146 NaN rows in hashtags. It is possible that this is simply due to the fact that some posts do not have hashtags.

In [7]:
data.isna().groupby('Caption')['Followings'].count()

Caption
False    24717
True        35
Name: Followings, dtype: int64

##### There are 35 NaN rows in Caption

### Maybe the caption or hashtag information is missing because there is either only a caption or only hashtags....

In [19]:
hashtag_nan = list(data['Hashtags'].isna())
caption_nan = list(data['Caption'].isna())

In [20]:
count = 0
for i in range(len(hashtag_nan)):
    if hashtag_nan[i] and caption_nan[i]:
        count += 1
print('The number of times both hashtag_nan and caption_nan is true is', count)

The number of times both hashtag_nan and caption_nan is true is 18


##### So we have missing hashtag both 18/35 times that we have a missing caption. So it is not exactly correlated. Maybe we can just drop the rows that have missing caption or hashtag if we need to use these, or claim that there was not hashtag or caption for these rows.

In [23]:
chi2_contingency([[hashtag_nan.count(True), hashtag_nan.count(False)], [caption_nan.count(True), caption_nan.count(False)]])

(67.09615038449334, 2.585828482081104e-16, 1, array([[   90.5, 24661.5],
        [   90.5, 24661.5]]))

### I want to try and see if I can make brand category numeric

In [24]:
brand_category = pd.get_dummies(data['BrandCategory'])

In [25]:
brand_category.head()

Designer  High street  Mega couture  \
UserId                                                                
1171579752855683619_212070047          0            1             0   
1171594777274371222_176762322          0            1             0   
1171407552643586413_581125501          0            1             0   
1171407552643586413_581125501          0            1             0   
1171508187966229230_2797323089         0            1             0   

                                Small couture  
UserId                                         
1171579752855683619_212070047               0  
1171594777274371222_176762322               0  
1171407552643586413_581125501               0  
1171407552643586413_581125501               0  
1171508187966229230_2797323089              0

In [28]:
data_new = pd.concat([data, brand_category], sort = False, axis = 1)
data_new.head()

Followings  Followers  MediaCount  \
UserId                                                              
1171579752855683619_212070047        518.0     9840.0         541   
1171594777274371222_176762322       7333.0     2300.0         272   
1171407552643586413_581125501        131.0      605.0         106   
1171407552643586413_581125501        131.0      605.0         106   
1171508187966229230_2797323089       132.0      456.0          56   

                                  BrandName BrandCategory  \
UserId                                                      
1171579752855683619_212070047   abercrombie   High street   
1171594777274371222_176762322   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171508187966229230_2797323089  abercrombie   High street   

                                                                         Hashtags  \
UserId                                                                              
1171579752855683619_212070047   beautiful, summer, fashion, love, cute, food, ...   
1171594777274371222_176762322   teen, model, brunette, selfie, hollister, snap...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171508187966229230_2797323089  abercrombieandfitch, shopping, love, hollister...   

                                                                          Caption  \
UserId                                                                              
1171579752855683619_212070047     We were born to be REAL, not to be perfect. ...   
1171594777274371222_176762322     Gotta run, but first, let me take selfie. Me...   
1171407552643586413_581125501                         dress to impress\U0001f457    
1171407552643586413_581125501                         dress to impress\U0001f457    
1171508187966229230_2797323089    #love #shopping #shoppen #hollister #abercro...   

                                                                           ImgURL  \
UserId                                                                              
1171579752855683619_212070047   https://scontent.cdninstagram.com/t51.2885-15/...   
1171594777274371222_176762322   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171508187966229230_2797323089  https://scontent.cdninstagram.com/t51.2885-15/...   

                                Likes  Comments   CreationTime  \
UserId                                                           
1171579752855683619_212070047      97          0    1453883211   
1171594777274371222_176762322      94          0    1453885002   
1171407552643586413_581125501      91          3    1453862683   
1171407552643586413_581125501      94          3    1453862683   
1171508187966229230_2797323089      9          1    1453874680   

                                                                    Link  \
UserId                                                                     
1171579752855683619_212070047   https://www.instagram.com/p/BBCSZPzAMIj/   
1171594777274371222_176762322   https://www.instagram.com/p/BBCVz4YJEyW/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171508187966229230_2797323089  https://www.instagram.com/p/BBCCH1zMEru/   

                                  Selfie  BodySnap  Marketing  ProductOnly  \
UserId                                                                       
1171579752855683619_212070047   0.000885  0.336438   0.741220     0.009454   
1171594777274371222_176762322   0.003912  0.995503   0.001728     0.002125   
1171407552643586413_5811

##### Now the brand category is represented in the last four columns of the dataframe with one_hot_encoding.

### In order to model with these we need to normalize the variables

# Model with Numeric Data Only

First, lets try to build a model with only the numeric data. This will be easy to do and will allow us to understand whether any of these numeric factors influence the number of likes/comments a user gets. 

### Likes Model

In [29]:
X = data_new[['Selfie', 'BodySnap', 'Marketing', 'ProductOnly', 'NonFashion',
       'Face', 'Logo', 'BrandLogo', 'Smile', 'Outdoor', 'NumberOfPeople',
       'NumberOfFashionProduct', 'Anger', 'Contempt', 'Disgust', 'Fear',
       'Happiness', 'Neutral', 'Sadness', 'Surprise', 'Designer',
       'High street', 'Mega couture', 'Small couture']]
y = data_new['Likes']

In [30]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [31]:
def standardize(train, test):


    mean = np.mean(train, axis=0)
    std = np.std(train, axis=0)+0.000001

    X_train = (train - mean) / std
    X_test = (test - mean) /std
    return X_train, X_test

In [32]:
X_train, X_test = standardize(X_train, X_test)

Let's use a Lasso Regression model to regress based on the likes that we receive. I want to use lasso because lasso will send the unimportant variables to zero allowing us to see which features are important. 

In [61]:
lasso = Lasso(alpha = 0.1, max_iter = 10000).fit(X_train, y_train)

In [62]:
predictions = lasso.predict((X_test))
r2_score(y_test, predictions)

0.012720167729573517

Maybe ridge regression will work better?

In [65]:
ridge = Ridge(alpha = 0.1).fit(X_train, y_train)

In [66]:
predictions = ridge.predict((X_test))
r2_score(y_test, predictions)

0.01269025248212674

### Including things we know likes are dependent on 

In [67]:
X = data_new[['Followings', 'Followers', 'MediaCount','Selfie', 'BodySnap', 'Marketing', 'ProductOnly', 'NonFashion',
       'Face', 'Logo', 'BrandLogo', 'Smile', 'Outdoor', 'NumberOfPeople',
       'NumberOfFashionProduct', 'Anger', 'Contempt', 'Disgust', 'Fear',
       'Happiness', 'Neutral', 'Sadness', 'Surprise', 'Designer',
       'High street', 'Mega couture', 'Small couture']]
y = data_new['Likes']

In [68]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [69]:
def standardize(train, test):


    mean = np.mean(train, axis=0)
    std = np.std(train, axis=0)+0.000001

    X_train = (train - mean) / std
    X_test = (test - mean) /std
    return X_train, X_test

In [70]:
X_train, X_test = standardize(X_train, X_test)

Let's use a Lasso Regression model to regress based on the likes that we receive. I want to use lasso because lasso will send the unimportant variables to zero allowing us to see which features are important. 

In [71]:
lasso = Lasso(alpha = 0.1, max_iter = 10000).fit(X_train, y_train)

In [72]:
predictions = lasso.predict((X_test))
r2_score(y_test, predictions)

0.2984377062815089

Maybe ridge regression will work better?

In [73]:
ridge = Ridge(alpha = 0.1).fit(X_train, y_train)

In [74]:
predictions = ridge.predict((X_test))
r2_score(y_test, predictions)

0.2984484171726812

So ridge regression is really no different...

### Try a Deep Learning Model

In [171]:
def our_model():
    model = Sequential()

    model.add(Dense(units=64, activation='relu', input_dim = 24))
    model.add(Dense(units=64, activation='relu'))
    model.add(Dense(units=64, activation='relu'))
    model.add(Dense(units=1))
    model.compile(loss='mean_squared_error', optimizer='adam',metrics=['accuracy'])
    return model

In [180]:
estimator = KerasRegressor(build_fn=our_model, epochs=100, batch_size=5, verbose=1)

In [169]:
from sklearn.model_selection import KFold, cross_val_score

kfold = KFold(n_splits=3)
results = cross_val_score(estimator, X_train, y_train, cv=kfold)
print("Baseline: %.2f (%.2f) MSE" % (results.mean(), results.std()))

/Users/simi/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:530: FutureWarning: From version 0.22, errors during fit will result in a cross validation score of NaN by default. Use error_score='raise' if you want an exception raised or error_score=np.nan to adopt the behavior from version 0.22.
  FutureWarning)


ValueError: Error when checking input: expected dense_122_input to have shape (27,) but got array with shape (24,)

In [173]:
estimator.fit(X_train, y_train)

Epoch 1/100
16583/16583 [==============================] - 3s 163us/step - loss: 947022.7400 - accuracy: 0.0025
Epoch 2/100
16583/16583 [==============================] - 3s 152us/step - loss: 946287.0022 - accuracy: 0.0014
Epoch 3/100
16583/16583 [==============================] - 3s 158us/step - loss: 946338.0476 - accuracy: 0.0016
Epoch 4/100
16583/16583 [==============================] - 3s 154us/step - loss: 945856.4013 - accuracy: 0.0017
Epoch 5/100
16583/16583 [==============================] - 3s 153us/step - loss: 945696.0217 - accuracy: 0.0011
Epoch 6/100
16583/16583 [==============================] - 3s 153us/step - loss: 945359.7348 - accuracy: 0.0014
Epoch 7/100
16583/16583 [==============================] - 3s 153us/step - loss: 945121.1274 - accuracy: 0.0018
Epoch 8/100
16583/16583 [==============================] - 3s 153us/step - loss: 945184.4540 - accuracy: 0.0015
Epoch 9/100
16583/16583 [==============================] - 3s 153us/step - loss: 944917.5923 - accuracy:

In [174]:
preds = estimator.predict(X_test)

8169/8169 [==============================] - 0s 33us/step


In [175]:
r2_score(y_test, preds)

0.01594369857292377

Maybe for the neural network model to work we need to normalize the outputs as well??? I have only normalized the input space. Maybe we need to investigate how the input space was normalized. Maybe the categorical variables that are represented by 1's and 0's at the end of the dataframe need to be normalized?

Average loss value for Keras deep learning model based on just numeric values and predicting likes is 154K.

Maybe try using likes / followers?

# Likes to Follower Ratio

First check if there are any null likes or follower entries

In [19]:
data_new['Likes'].isnull().any() 

False

In [20]:
data_new['Followers'].isnull().any()

False

In [21]:
data_ver2 = pd.concat([data, brand_category], sort = False, axis = 1)
data_ver2['LikesToFollowers'] = data_ver2['Likes'] / data_ver2['Followers']
data_ver2.head()

Followings  Followers  MediaCount  \
UserId                                                              
1171579752855683619_212070047        518.0     9840.0         541   
1171594777274371222_176762322       7333.0     2300.0         272   
1171407552643586413_581125501        131.0      605.0         106   
1171407552643586413_581125501        131.0      605.0         106   
1171508187966229230_2797323089       132.0      456.0          56   

                                  BrandName BrandCategory  \
UserId                                                      
1171579752855683619_212070047   abercrombie   High street   
1171594777274371222_176762322   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171508187966229230_2797323089  abercrombie   High street   

                                                                         Hashtags  \
UserId                                                                              
1171579752855683619_212070047   beautiful, summer, fashion, love, cute, food, ...   
1171594777274371222_176762322   teen, model, brunette, selfie, hollister, snap...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171508187966229230_2797323089  abercrombieandfitch, shopping, love, hollister...   

                                                                          Caption  \
UserId                                                                              
1171579752855683619_212070047     We were born to be REAL, not to be perfect. ...   
1171594777274371222_176762322     Gotta run, but first, let me take selfie. Me...   
1171407552643586413_581125501                         dress to impress\U0001f457    
1171407552643586413_581125501                         dress to impress\U0001f457    
1171508187966229230_2797323089    #love #shopping #shoppen #hollister #abercro...   

                                                                           ImgURL  \
UserId                                                                              
1171579752855683619_212070047   https://scontent.cdninstagram.com/t51.2885-15/...   
1171594777274371222_176762322   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171508187966229230_2797323089  https://scontent.cdninstagram.com/t51.2885-15/...   

                                Likes  Comments   CreationTime  \
UserId                                                           
1171579752855683619_212070047      97          0    1453883211   
1171594777274371222_176762322      94          0    1453885002   
1171407552643586413_581125501      91          3    1453862683   
1171407552643586413_581125501      94          3    1453862683   
1171508187966229230_2797323089      9          1    1453874680   

                                                                    Link  \
UserId                                                                     
1171579752855683619_212070047   https://www.instagram.com/p/BBCSZPzAMIj/   
1171594777274371222_176762322   https://www.instagram.com/p/BBCVz4YJEyW/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171508187966229230_2797323089  https://www.instagram.com/p/BBCCH1zMEru/   

                                  Selfie  BodySnap  Marketing  ProductOnly  \
UserId                                                                       
1171579752855683619_212070047   0.000885  0.336438   0.741220     0.009454   
1171594777274371222_176762322   0.003912  0.995503   0.001728     0.002125   
1171407552643586413_5811

In [22]:
data_ver2.replace([np.inf, -np.inf], np.nan, inplace = True);

In [23]:
data_test = data_ver2[data_ver2['LikesToFollowers'].isnull()]
data_test.head()

Followings  Followers  MediaCount  \
UserId                                                              
1171240019544303946_23457617           0.0        0.0           1   
1171142669094830425_252313279         36.0        0.0           0   
1171278619918133139_23638768           0.0        0.0           0   
1170503373243750859_2795691477         0.0        0.0           1   
1170007750938083500_2120361935        39.0        0.0           0   

                                    BrandName BrandCategory  \
UserId                                                        
1171240019544303946_23457617      acnestudios      Designer   
1171142669094830425_252313279     acnestudios      Designer   
1171278619918133139_23638768      acnestudios      Designer   
1170503373243750859_2795691477    acnestudios      Designer   
1170007750938083500_2120361935  alexanderwang      Designer   

                                                                         Hashtags  \
UserId                                                                              
1171240019544303946_23457617    shop, anndemeulemeester, style, acnestudios, t...   
1171142669094830425_252313279                            acnestudios, collegeface   
1171278619918133139_23638768                                                  NaN   
1170503373243750859_2795691477  sorpreseinmacchina, acnestudios, withstyle, ga...   
1170007750938083500_2120361935  jimmychoo, slimthickwithyourcuteass, alexander...   

                                                                          Caption  \
UserId                                                                              
1171240019544303946_23457617      Black? Or dark black... @00000 #totokaelo #l...   
1171142669094830425_252313279     #collegeface in college\U0001f610 #acnestudios    
1171278619918133139_23638768                1.26 | \u2603  Cc: \U0001f4f7 @00000    
1170503373243750859_2795691477    #jackolsen #jauria #withstyle #migraineprobl...   
1170007750938083500_2120361935    u\"\"I swear she's so perfect she makes me s...   

                                                                           ImgURL  \
UserId                                                                              
1171240019544303946_23457617    https://scontent.cdninstagram.com/t51.2885-15/...   
1171142669094830425_252313279   https://scontent.cdninstagram.com/t51.2885-15/...   
1171278619918133139_23638768    https://scontent.cdninstagram.com/t51.2885-15/...   
1170503373243750859_2795691477  https://scontent.cdninstagram.com/t51.2885-15/...   
1170007750938083500_2120361935  https://scontent.cdninstagram.com/t51.2885-15/...   

                                Likes  Comments   CreationTime  \
UserId                                                           
1171240019544303946_23457617       37          0    1453842712   
1171142669094830425_252313279      18          1    1453831107   
1171278619918133139_23638768      136          9    1453847313   
1170503373243750859_2795691477     10          0    1453754897   
1170007750938083500_2120361935     74          0    1453695814   

                                                                    Link  \
UserId                                                                     
1171240019544303946_23457617    https://www.instagram.com/p/BBBFJefqOlK/   
1171142669094830425_252313279   https://www.instagram.com/p/BBAvA10s0FZ/   
1171278619918133139_23638768    https://www.instagram.com/p/BBBN7L5OueT/   
1170503373243750859_2795691477  https://www.instagram.com/p/BA-dp3KlxnL/   
1170007750938083500_2120361935  https://www.instagram.com/p/BA8s9m7HkCs/   

                                  Selfie  BodySnap  Marketing  ProductOnly  \
UserId                                                                       
1171240019544303946_23457617    0.000158  0.001617   0.007038     0.997941   
1171142669094830425_252313279   0.870760  0.200230   0.063385     0.003886   
1171278619

Drop rows where LikesToFollowers is NAN. Likes and Followers count at both 0.

In [24]:
data_ver2 = data_ver2.dropna()

In [25]:
data_ver2['LikesToFollowers'].isnull().any()

False

In [26]:
X = data_ver2[['Selfie', 'BodySnap', 'Marketing', 'ProductOnly', 'NonFashion',
       'Face', 'Logo', 'BrandLogo', 'Smile', 'Outdoor', 'NumberOfPeople',
       'NumberOfFashionProduct', 'Anger', 'Contempt', 'Disgust', 'Fear',
       'Happiness', 'Neutral', 'Sadness', 'Surprise', 'Designer',
       'High street', 'Mega couture', 'Small couture']]
y = data_ver2['LikesToFollowers']

In [27]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [28]:
def standardize(train, test):


    mean = np.mean(train, axis=0)
    std = np.std(train, axis=0)+0.000001

    X_train = (train - mean) / std
    X_test = (test - mean) /std
    return X_train, X_test

In [29]:
X_train, X_test = standardize(X_train, X_test)

In [30]:
X_train

Selfie  BodySnap  Marketing  ProductOnly  \
UserId                                                                       
1171507821618345730_1574502508 -0.259308 -0.824482  -0.528553     1.122620   
1171231788214163629_193097469  -0.257499 -0.821518  -0.395204     1.372306   
1170757716739212431_33474720   -0.083411  1.660241  -0.434811    -0.930392   
1160779262971561635_1558220976 -0.235183  0.118235  -0.324666    -0.438253   
1171512225661878713_1643536695  0.009773 -0.367641   0.564813    -0.880679   
...                                  ...       ...        ...          ...   
1171371918978904970_1255231723 -0.128482  1.514154  -0.311073    -0.922707   
1101082779343622525_498541251  -0.245017  1.223258  -0.471399    -0.644079   
1171555437158436633_205040098  -0.257739 -0.819451  -0.507412     1.364928   
1169853176213918638_1532376523  5.020211 -0.143800  -0.077522    -0.921410   
1170368427511165568_1302219    -0.251476 -0.804806  -0.528797    -0.094964   

                                NonFashion      Face      Logo  BrandLogo  \
UserId                                                                      
1171507821618345730_1574502508    0.706286 -0.819929  1.338838  -0.632729   
1171231788214163629_193097469    -0.469912 -0.817212  0.905760   0.096165   
1170757716739212431_33474720     -0.429279  1.613308 -0.885282  -0.690709   
1160779262971561635_1558220976   -0.401616 -0.640163 -1.157490  -0.588286   
1171512225661878713_1643536695    0.103549  0.873789 -0.031832  -0.240964   
...                                    ...       ...       ...        ...   
1171371918978904970_1255231723   -0.287621  1.267836 -0.102080  -0.339481   
1101082779343622525_498541251    -0.470364 -0.758604 -1.276440  -0.775869   
1171555437158436633_205040098    -0.465846 -0.818924 -0.843107  -0.214763   
1169853176213918638_1532376523   -0.338819  1.562264 -0.969135  -0.541927   
1170368427511165568_1302219       2.541495 -0.813678 -0.858515  -0.674339   

                                   Smile   Outdoor  NumberOfPeople  \
UserId                                                               
1171507821618345730_1574502508 -0.592633 -0.499348       -0.957247   
1171231788214163629_193097469  -0.589268 -0.490598       -0.765123   
1170757716739212431_33474720    3.174731  0.134557        2.035478   
1160779262971561635_1558220976 -0.513935  0.260650        0.004726   
1171512225661878713_1643536695  0.174241 -0.148754        0.662870   
...                                  ...       ...             ...   
1171371918978904970_1255231723  0.785832  0.174616        1.066442   
1101082779343622525_498541251  -0.543664 -0.432439       -0.455082   
1171555437158436633_205040098  -0.591105 -0.451280       -0.836088   
1169853176213918638_1532376523  1.939863 -0.419760        2.027119   
1170368427511165568_1302219    -0.580670 -0.497198       -0.830845   

                                NumberOfFashionProduct     Anger  Contempt  \
UserId                                                                       
1171507821618345730_1574502508               -1.366377 -0.102171 -0.123572   
1171231788214163629_193097469                 0.273672 -0.102171 -0.123572   
1170757716739212431_33474720                  1.229400  0.048616 -0.028745   
1160779262971561635_1558220976                0.557519 -0.102171 -0.123572   
1171512225661878713_1643536695               -0.334646  0.461866  0.228294   
...                                                ...       ...       ...   
1171371918978904970_1255231723                0.270878  0.014871 -0.075847   
1101082779343622525_498541251                 0.531988 -0.102171 -0.123572   
1171555437158436633_205040098                -0.276188 -0.102171 -0.123572   
1169853176213918638_1532376523               -0.731254 -0.102171 -0.123572   
1170368427511165568_1302219                  -1.240286 -0.102171 -0.123572   

                                 Disgust      Fear  Happiness   Neutral  \
UserId          

In [31]:
import keras
keras.backend.floatx()
keras.backend.set_floatx('float64')
keras.backend.floatx()

'float64'

In [126]:
def our_model_ver2():
    model = Sequential()
    model.add(Dense(units=128, input_dim = 24))
    model.add(LeakyReLU(alpha=0.1))
    model.add(Dropout(0.25))

    model.add(Dense(units=256))
    model.add(LeakyReLU(alpha=0.1))
    model.add(Dropout(0.25))
    
    model.add(Dense(units=64, activation = 'tanh'))
    
    model.add(Dense(units=1))
    #opt = Adadelta(clipnorm=0.3)
    model.compile(loss='mean_squared_error', optimizer='adam',metrics=['accuracy'])
    return model

estimator = KerasRegressor(build_fn=our_model_ver2, epochs=100, batch_size=5000, verbose=1)

# kfold = KFold(n_splits=5)
# results = cross_val_score(estimator, X_train, y_train, cv=kfold)
# print("Baseline: %.2f (%.2f) MSE" % (results.mean(), results.std()))

In [127]:
estimator.fit(X_train, y_train)

Epoch 1/100
16443/16443 [==============================] - 0s 21us/step - loss: 21.2166 - accuracy: 0.0140
Epoch 2/100
16443/16443 [==============================] - 0s 11us/step - loss: 21.1206 - accuracy: 0.0136
Epoch 3/100
16443/16443 [==============================] - 0s 11us/step - loss: 21.0924 - accuracy: 0.0108
Epoch 4/100
16443/16443 [==============================] - 0s 11us/step - loss: 21.0914 - accuracy: 0.0138
Epoch 5/100
16443/16443 [==============================] - 0s 11us/step - loss: 21.0649 - accuracy: 0.0141
Epoch 6/100
16443/16443 [==============================] - 0s 11us/step - loss: 21.0362 - accuracy: 0.0128
Epoch 7/100
16443/16443 [==============================] - 0s 11us/step - loss: 21.0704 - accuracy: 0.0139
Epoch 8/100
16443/16443 [==============================] - 0s 11us/step - loss: 21.0444 - accuracy: 0.0145
Epoch 9/100
16443/16443 [==============================] - 0s 11us/step - loss: 21.0408 - accuracy: 0.0136
Epoch 10/100
16443/16443 [===========

In [128]:
preds = estimator.predict(X_test)

8100/8100 [==============================] - 0s 7us/step


In [129]:
r2_score(y_test, preds)

0.020305390606470408